File: src/css/properties/properties_impl.zig
const bun = @import("bun");

pub const css = @import("../css_parser.zig");

const CustomPropertyName = css.css_properties.CustomPropertyName;

const Printer = css.Printer;
const PrintErr = css.PrintErr;
const VendorPrefix = css.VendorPrefix;

const PropertyId = css.PropertyId;
const Property = css.Property;

pub const property_id_mixin = struct {
    pub fn toCss(this: *const PropertyId, comptime W: type, dest: *Printer(W)) PrintErr!void {
        var first = true;
        const name = this.name();
        const prefix_value = this.prefix().orNone();

        inline for (VendorPrefix.FIELDS) |field| {
            if (@field(prefix_value, field)) {
                var prefix: VendorPrefix = .{};
                @field(prefix, field) = true;

                if (first) {
                    first = false;
                } else {
                    try dest.delim(',', false);
                }
                try prefix.toCss(W, dest);
                try dest.writeStr(name);
            }
        }
    }

    pub fn parse(input: *css.Parser) css.Result(PropertyId) {
        const name = switch (input.expectIdent()) {
            .result => |v| v,
            .err => |e| return .{ .err = e },
        };
        return .{ .result = fromString(name) };
    }

    pub fn fromString(name_: []const u8) PropertyId {
        const name_ref = name_;
        var prefix: VendorPrefix = undefined;
        var trimmed_name: []const u8 = undefined;

        // TODO: todo_stuff.match_ignore_ascii_case
        if (bun.strings.startsWithCaseInsensitiveAscii(name_ref, "-webkit-")) {
            prefix = VendorPrefix{ .webkit = true };
            trimmed_name = name_ref[8..];
        } else if (bun.strings.startsWithCaseInsensitiveAscii(name_ref, "-moz-")) {
            prefix = VendorPrefix{ .moz = true };
            trimmed_name = name_ref[5..];
        } else if (bun.strings.startsWithCaseInsensitiveAscii(name_ref, "-o-")) {
            prefix = VendorPrefix{ .o = true };
            trimmed_name = name_ref[3..];
        } else if (bun.strings.startsWithCaseInsensitiveAscii(name_ref, "-ms-")) {
            prefix = VendorPrefix{ .ms = true };
            trimmed_name = name_ref[4..];
        } else {
            prefix = VendorPrefix{ .none = true };
            trimmed_name = name_ref;
        }

        return PropertyId.fromNameAndPrefix(trimmed_name, prefix) orelse .{ .custom = CustomPropertyName.fromStr(name_) };
    }
};

pub const property_mixin = struct {
    /// Serializes the CSS property, with an optional `!important` flag.
    pub fn toCss(this: *const Property, comptime W: type, dest: *Printer(W), important: bool) PrintErr!void {
        if (this.* == .custom) {
            try this.custom.name.toCss(W, dest);
            try dest.delim(':', false);
            try this.valueToCss(W, dest);
            if (important) {
                try dest.whitespace();
                try dest.writeStr("!important");
            }
            return;
        }
        const name, const prefix = this.__toCssHelper();
        var first = true;

        inline for (VendorPrefix.FIELDS) |field| {
            if (@field(prefix, field)) {
                var p: VendorPrefix = .{};
                @field(p, field) = true;

                if (first) {
                    first = false;
                } else {
                    try dest.writeChar(';');
                    try dest.newline();
                }
                try p.toCss(W, dest);
                try dest.writeStr(name);
                try dest.delim(':', false);
                try this.valueToCss(W, dest);
                if (important) {
                    try dest.whitespace();
                    try dest.writeStr("!important");
                }
            }
        }
    }
};

File: src/css/properties/properties_impl.rs
use crate as css;

use css::PrintErr;
use css::Printer;
use css::VendorPrefix;
use css::css_properties::CustomPropertyName;
use css::css_properties::{Property, PropertyId, PropertyIdTag};

impl Property {
    #[inline]
    pub fn variant_tag(&self) -> PropertyIdTag {
        match self {
            Property::Unparsed(_) => PropertyIdTag::Unparsed,
            Property::Custom(_) => PropertyIdTag::Custom,
            // Every other `Property` variant maps 1:1 onto its `PropertyId`
            // variant, so `property_id().tag()` is the discriminant.
            _ => self.property_id().tag(),
        }
    }
}

pub(super) const PREFIX_FLAGS: [VendorPrefix; 5] = [
    VendorPrefix::WEBKIT,
    VendorPrefix::MOZ,
    VendorPrefix::MS,
    VendorPrefix::O,
    VendorPrefix::NONE,
];

pub(super) mod property_id_mixin {
    use super::*;

    pub(crate) fn to_css(this: &PropertyId, dest: &mut Printer) -> Result<(), PrintErr> {
        let name = this.name();
        let prefix_value = this.prefix().or_none();

        // PORT NOTE: Zig `inline for (VendorPrefix.FIELDS) |field|` + `@field` iterates each
        // bitflag field and tests it. `PREFIX_FLAGS` is the same set in the same order;
        // `contains` replaces the `@field` test.
        dest.write_comma_separated(
            PREFIX_FLAGS
                .iter()
                .copied()
                .filter(|p| prefix_value.contains(*p)),
            |d, p| {
                p.to_css(d)?;
                d.write_str(name)
            },
        )
    }

    pub(crate) fn parse(input: &mut css::Parser) -> css::Result<PropertyId> {
        // PORT NOTE: `css::Result<T>` is assumed to alias `Result<T, css::ParserError>`;
        // the Zig `.result`/`.err` switch collapses to `?`.
        let name = input.expect_ident()?;
        Ok(from_string(name))
    }

    pub(crate) fn from_string(name_: &[u8]) -> PropertyId {
        let (prefix, trimmed_name) = VendorPrefix::strip_from(name_);
        PropertyId::from_name_and_prefix(trimmed_name, prefix)
            .unwrap_or_else(|| PropertyId::Custom(CustomPropertyName::from_str(name_)))
    }
}

pub(super) mod property_mixin {
    use super::*;

    /// Serializes the CSS property, with an optional `!important` flag.
    pub(crate) fn to_css(
        this: &Property,
        dest: &mut Printer,
        important: bool,
    ) -> Result<(), PrintErr> {
        if let Property::Custom(custom) = this {
            custom.name.to_css(dest)?;
            dest.delim(b':', false)?;
            this.value_to_css(dest)?;
            if important {
                dest.whitespace()?;
                dest.write_str(b"!important")?;
            }
            return Ok(());
        }
        let (name, prefix) = this.__to_css_helper();

        // PORT NOTE: see property_id_mixin::to_css for the `inline for` + `@field` mapping.
        dest.write_separated(
            PREFIX_FLAGS.iter().copied().filter(|p| prefix.contains(*p)),
            |d| {
                d.write_char(b';')?;
                d.newline()
            },
            |d, p| {
                p.to_css(d)?;
                d.write_str(name)?;
                d.delim(b':', false)?;
                this.value_to_css(d)?;
                if important {
                    d.whitespace()?;
                    d.write_str(b"!important")?;
                }
                Ok(())
            },
        )
    }
}

// ported from: src/css/properties/properties_impl.zig
